### PR TITLE
feat: refactor wallet logic

### PR DESCRIPTION
**Pull request by hstove**
> Try out this version of the Stacks Wallet - download [extension builds](https://github.com/blockstack/ux/actions/runs/479228698), the [hosted version](https://pr-758.app.stacks.engineering), or the [Stacks testnet demo app](https://pr-758.testnet-demo.stacks.engineering)

This PR is ultimately to fix the issue where your Stacks address would not match the address you'd get in the desktop wallet, or on other systems.

However, when I was building that logic, I went forward with a larger approach of refactoring. I didn't like a lot of the design of `@stacks/keychain` - it was built before we had anything in Stacks 2.0, and the class-based (OOP) design made it annoying to serialize, lock, encrypt, etc.

So, alas, a new library - `@stacks/wallet-sdk` - introduced just now in https://github.com/blockstack/stacks.js/pull/897 .

Not to bury the lead, but this PR also updates and removes some other old libraries:

- `@blockstack/stacks-transactions` -> `@stacks/transactions`
- Fully removed `blockstack.js`!!! 🥳 That includes in Connect. So, @friedger , we can close #743. It's been replaced with the latest relevant stack.js package.

### CODE REVIEW

**Comment by markmhendrickson** [comment]
@hstove should we assign a couple of code reviewers here?

File: packages/app/src/common/hooks/use-wallet.ts [hunk 1]
```diff
@@ -176,29 +174,28 @@ export const useWallet = () => {
           name: appName as string,
         },
       });
-      const { defaultUsername } = currentIdentity;
+      const { username } = account;
       if (NODE_ENV === 'test') {
-        currentIdentity.defaultUsername = '';
+        account.username = '';
```
Comment (aulneau) [inline packages/app/src/common/hooks/use-wallet.ts:179]: :C mutation

File: packages/app/src/common/hooks/use-wallet.ts [hunk 2]
```diff
@@ -176,29 +174,28 @@ export const useWallet = () => {
...
       }
-      const authResponse = await currentIdentity.makeAuthResponse({
-        gaiaUrl,
+      const authResponse = await makeAuthResponse({
+        gaiaHubUrl: gaiaUrl,
```
Comment (aulneau) [inline packages/app/src/common/hooks/use-wallet.ts:182]: name them both the same? `gaiaUrl -> gaiaHubUrl`

File: packages/app/src/common/hooks/use-wallet.ts [hunk 3]
```diff
@@ -176,29 +174,28 @@ export const useWallet = () => {
...
         appDomain: appURL.origin,
         transitPublicKey: decodedAuthRequest.public_keys[0],
         scopes: decodedAuthRequest.scopes,
+        account,
       });
-      currentIdentity.defaultUsername = defaultUsername;
+      account.username = username;
```
Comment (aulneau) [inline packages/app/src/common/hooks/use-wallet.ts:188]: why the mutation?

File: packages/app/src/common/hooks/use-wallet.ts [hunk 4]
```diff
@@ -210,7 +207,7 @@ export const useWallet = () => {
         appIcon = appManifest.icons[0].src as string;
       }
 
-      const hasIdentities = identities && identities.length;
+      const hasIdentities = wallet?.accounts && wallet.accounts.length;
```
Comment (aulneau) [inline packages/app/src/common/hooks/use-wallet.ts:210]: `wallet?.accounts?.length > 0`

File: packages/app/src/common/transaction-utils.ts [hunk 5]
```diff
@@ -31,26 +34,25 @@ const getPostConditions = (postConditions?: PostCondition[]): PostCondition[] |
...
   nonce,
 }: {
   txData: ContractCallPayload;
-  signer: WalletSigner;
+  senderKey: string;
   nonce?: number;
 }) => {
   const { contractName, contractAddress, functionName, functionArgs } = txData;
-  const version = TransactionVersion.Testnet;
   const args = functionArgs.map(arg => {
```
Comment (aulneau) [inline packages/app/src/common/transaction-utils.ts:45]: `args` -> `functionArgs`

File: packages/app/src/components/drawer/accounts/add-username.tsx [hunk 6]
```diff
@@ -1,42 +1,31 @@
 import React, { useState } from 'react';
 import { useWallet } from '@common/hooks/use-wallet';
 import { Box, Text, Input, Button, Flex } from '@stacks/ui';
-import { IdentityNameValidityError, registerSubdomain, validateSubdomain } from '@stacks/keychain';
-import { errorTextMap } from '@pages/username';
-import { ErrorLabel } from '@components/error-label';
+// import { ErrorLabel } from '@components/error-label';
```
Comment (aulneau) [inline packages/app/src/components/drawer/accounts/add-username.tsx:4]: remove comment

File: packages/app/src/components/drawer/confirm-send-drawer.tsx [hunk 7]
```diff
@@ -40,12 +40,17 @@ export const ConfirmSendDrawer: React.FC<ConfirmSendDrawerProps> = ({
...
+    doSetLatestNonce,
+    currentAccountStxAddress,
+  } = useWallet();
   const asset = useRecoilValue(selectedAssetStore);
   const [loading, setLoading] = useState(false);
   const { doChangeScreen } = useAnalytics();
   const getTx = useCallback(async () => {
-    if (!asset || !currentIdentity) return;
+    if (!asset || !currentAccount || !currentAccountStxAddress) return;
     const network = new StacksTestnet();
```
Comment (aulneau) [inline packages/app/src/components/drawer/confirm-send-drawer.tsx:54]: should this be hardcoded to the testnet?

File: packages/app/src/components/popup/settings-popover.tsx [hunk 8]
```diff
@@ -79,7 +79,7 @@ export const SettingsPopover: React.FC = () => {
       background="white"
       display={showing ? 'block' : 'none'}
     >
-      {identities && identities.length > 1 ? (
+      {wallet && wallet.accounts.length > 1 ? (
```
Comment (aulneau) [inline packages/app/src/components/popup/settings-popover.tsx:82]: `wallet?.accounts?.length`

File: packages/app/src/store/recoil/api.ts [hunk 9]
```diff
@@ -94,11 +98,11 @@ export const accountDataStore = selector({
   key: 'api.account-data',
   get: async ({ get }) => {
     const { url } = get(currentNetworkStore);
-    const currentIdentity = get(currentIdentityStore);
-    const principal = currentIdentity?.getStxAddress();
-    if (!principal) {
+    const account = get(currentAccountStore);
+    if (!account) {
       throw 'Unable to get account info when logged out.';
```
Comment (aulneau) [inline packages/app/src/store/recoil/api.ts:103]: should it throw `new Error()`?

File: packages/app/src/store/recoil/wallet.ts [hunk 10]
```diff
@@ -83,14 +70,17 @@ export const latestNonceStore = selector({
...
     return nonce;
   },
 });
 
-export const currentIdentityIndexStore = atom<number | undefined>({
-  key: 'wallet.current-identity-index',
+export const currentAccountIndexStore = atom<number | undefined>({
+  key: 'wallet.current-account-index',
   default: undefined,
   effects_UNSTABLE: [localStorageEffect()],
```
Comment (aulneau) [inline packages/app/src/store/recoil/wallet.ts:85]: shouldn't this take a `key`?

**Review by aulneau — APPROVED** [review]

File: packages/app/src/common/hooks/use-wallet.ts (hunk 1 above)
Comment (hstove) [reply packages/app/src/common/hooks/use-wallet.ts:179]: Sorry, I'll fix this. It's only for testing, the integration tests have to fetch and validate usernames. This is bad, you're right.

File: packages/app/src/components/drawer/confirm-send-drawer.tsx (hunk 7 above)
Comment (hstove) [reply packages/app/src/components/drawer/confirm-send-drawer.tsx:54]: I don't have the same ChainID logic that you have in my network store. Once I have that, `currentAccountStxAddress` will be a recoil selector.

File: packages/app/src/store/recoil/wallet.ts (hunk 10 above)
Comment (hstove) [reply packages/app/src/store/recoil/wallet.ts:85]: It's a builder-style function... `localStorageEffect()` returns a function that accepts the atom as it's first param (in line with the recoil effects API.

**Comment by markmhendrickson** [comment]
This will go live with https://github.com/blockstack/ux/pull/761

**Pull request closed without merging**